### PR TITLE
Increase shards

### PR DESCRIPTION
**Pull request by benrudolph**
We may want to combine this with @calellowitz [PR](https://github.com/dimagi/commcare-hq/pull/15514), but I never love combining changes, so would rather not if we think it's doable to just run the AppSubmissionTracker pillow reindex on its own. I looked at our disk usage and I think we'll actually be fine not reallocating another node. Our most full disk still has around 120 GB of free space. hqcases is 61gb and xforms is 110gb. That will roughly add 60GB - 70GB to each node conservatively. 

1. `fab <softlayer|production> setup_release:keep_days=10 --set code_branch=br/more-shards`
1.5. disable shard reallocation
2. `./manage.py ptop_preindex --reset` (in a screen)
3. Merge this PR
4. Deploy
5. Run `./manage.py prune_elastic_indices` once we feel good that everything is settled

es head should then look something like this except distributed amongst 3 nodes
screen shot 2017-03-28 at 1 20 44 pm

### CODE REVIEW

**Comment by benrudolph** [comment]
arguably might be better to have 9 shards

**Comment by emord** [comment]
What if we made number of shards dependent on `SERVER_ENVIRONMENT`? I imagine we don't want the overhead on monoliths like swiss.

Also we should think about icds and thing about how many shards we expect to need

**Comment by benrudolph** [comment]
that's a good point, it's a little unfortunate that all of this gets set in the django code. i think it may be best to specify in the localsettings, what do you think?

**Comment by emord** [comment]
Yeah. I'm pro moving anything like that to localsettings

**Comment by benrudolph** [comment]
actually doing this in the localsettings makes it really hard to reindex in a private release

**Comment by benrudolph** [comment]
https://github.com/dimagi/commcare-hq/pull/15550/commits/60b6717d38bbbf83b01753b93d2bb46d8e936179 -- this makes me cringe, but if we move this to localsettings we have to make some infrastructure for ansible to be able to deploy to a private release which seems non-intuitive for changes in the future

**Comment by emord** [comment]
ouch. that's gross, but agreed we don't have a great way to do that with ansible yet.

**Review by emord — APPROVED** [review]
Looks good. Maybe it'd be worthwhile to set https://github.com/dimagi/commcare-hq/blob/d26d796c9c50edac166fb01bc83719381482e0ce/corehq/ex-submodules/pillowtop/es_utils.py#L10 to -1 and index it once the reindex finishes? or at least bump it past 15 min?

**Review by millerdev — COMMENTED** [review]
the code changes look fine, but I agree would be better to have this in env-specific configuration like localsettings. I didn't understand the above-mentioned problem with using localsettings, but that's probably because I don't have enough context on ES in general.

**Comment by benrudolph** [comment]
ok doubled it to 30 mins.

@millerdev it's hard with ansible since ansible always deploys to the `current` release. however we want to do this in a private release so we can build up the indices separately, and then when we merge this PR the switch over to the new indices will be fairly painless

File: corehq/pillows/mappings/case_mapping.py [hunk 1]
```diff
@@ -104,10 +106,19 @@
...
 CASE_ES_ALIAS = "hqcases"
 
+CASE_INDEX_META = deepcopy(DEFAULT_META)
+
+if settings.SERVER_ENVIRONMENT == 'production':
+    CASE_INDEX_META.update({
+        'settings': {
+            'number_of_shards': 10,
+        },
+    })
```
Comment (snopoke) [inline corehq/pillows/mappings/case_mapping.py]: could we make use of the `environments.yml` file for this kind of thing?

**Comment by snopoke** [comment]
```
======================================================================
FAIL: test_instantiate_all (testapps.test_pillowtop.tests.test_settings.PillowtopSettingsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/testapps/test_pillowtop/tests/test_settings.py", line 30, in test_instantiate_all
    self.assertEqual(expected_meta[pillow_config.name], _pillow_meta_from_config(pillow_config))
AssertionError: {u'change_feed_type': u'KafkaChangeFeed', u'full_class_name': u'pillowtop.pillow [truncated]... != {'change_feed_type': 'KafkaChangeFeed', 'checkpoint_id': u'CaseToElasticsearchPi [truncated]...
- {u'advertised_name': u'CaseToElasticsearchPillow',
?  -                   -
+ {'advertised_name': 'CaseToElasticsearchPillow',
-  u'change_feed_type': u'KafkaChangeFeed',
?  -                    -
+  'change_feed_type': 'KafkaChangeFeed',
-  u'checkpoint_id': u'CaseToElasticsearchPillow-test_hqcases_2016-03-04',
?  -                                                             ^    ^^
+  'checkpoint_id': u'CaseToElasticsearchPillow-test_hqcases_2017-03-28',
?                                                               ^    ^^
-  u'full_class_name': u'pillowtop.pillow.interface.ConstructedPillow',
?  -                   -
+  'full_class_name': 'pillowtop.pillow.interface.ConstructedPillow',
-  u'name': u'CaseToElasticsearchPillow'}
?  -        -
+  'name': 'CaseToElasticsearchPillow'}
```

**Comment by dannyroberts** [comment]
@benrudolph for those of us who weren't following closely on slack#pillows but want to understand, can you go over some basics such as

- how many shards there were before/are by default
- whether shards are a per-index thing (looks like they are)
- basic mental model I can use of what shards in ES are
- what we expect to happen by increasing the number of shards
- what the tradeoff is (i.e. why 10 is better than 1 but worse than 1000)

Thanks, and sorry to make you summarize, but hopefully the explanation will be useful to others as well.

File: corehq/pillows/mappings/case_mapping.py (hunk 1 above)
Comment (benrudolph) [inline corehq/pillows/mappings/case_mapping.py]: we could do that, i'm a little hesitant because then our ES settings would live in ansible, HQ, and commcare-hq-deploy. i was already confused initially about where to make this change

**Comment by benrudolph** [comment]
@dannyroberts sure

> - how many shards there were before/are by default


By default there are 5

> - whether shards are a per-index thing (looks like they are)


Yes

> - basic mental model I can use of what shards in ES are


Shards are under the hood Lucene indexes. Elasticsearch indexes are made up of many Lucene indexes in order to scale horizontally. Elasticsearch tries to balance the shards evenly across the number of nodes (servers). Currently we have 5 shards that is attempting to be balanced across 3 nodes, so it's a bit uneven at the moment.

> - what we expect to happen by increasing the number of shards


Two things. 1) a more even distribution of shards/shard size. Since shards are the ~same size, have 2 20GB shards and 1 node and 1 20GB shard on another is a fairly big difference. However if we have 3 shards at 10GB on one node and 4 shards of 10GB on another that's not as large as a difference. 2) We think that when ES tries to rebalance shards, the form and case shards are too big and causing it to fall over. We hope that with smaller shards it will be able to more easily to balance the shards

> - what the tradeoff is (i.e. why 10 is better than 1 but worse than 1000)


My understanding, which may not be totally correct, is that there is overhead to maintaining a larger number of shards and querying across all of them, so we don't want an outrageously high number if we can help. At the same time we want our shards to be evenly balanced amongst our nodes so that we can distribute the load more evenly, which is why 10 is better than 1.

Let me know if any of that makes (or @emord / @snopoke correct me if I've misunderstood something)

**Comment by snopoke** [comment]
That's true. I think localsettings is the right place for this to live with
our current workflows. The issue that you mentioned earlier is that we're
changing the value Pre deploy which doesn't really work right now.

If we changed the workflow to make Ansible create the indexes (instead of
HQ) then this wouldn't be an issue but it would mean an extra step anytime
we added an index which isn't very often.

**Comment by benrudolph** [comment]
I would be very pro ansible managing our ES indexes. it's weird to me as is that they are so closely tied to pillows

**Comment by emord** [comment]
One thing that might make that hard is that UCR data sources can create ES indexes. Although I think there could be a workaround for that if the datasources are refactored

**Comment by snopoke** [comment]
I was only thinking of the static indexes - fine to have UCR's dynamically
create indexes.

**Comment by dannyroberts** [comment]
Thanks @benrudolph https://github.com/dimagi/commcare-hq/pull/15550#issuecomment-290129280 was helpful

**Comment by emord** [comment]
@benrudolph not sure where your experimentation with different date formats left off, but if it seemed promising or maybe inconclusive due to too little data, it might be a good time to add some extra columns for formats on received on so we can profile

**Comment by benrudolph** [comment]
i stopped with the `epoch_second` (figured it wasn't the root cause to our pain, so stopped looking into it). agree would be fairly easy to add some new formats here. i'm a bit torn on this PR as i'm not sure how to move forward.

1. moving all of this to ansible feels out of scope for this change
2. move this configuration to commcare-hq-deploy environments.yml 
3. continue the hacky way forward
4. we create a `prod_settings.py` file (similar to `dev_settings.py`) where we configure all of this

**Comment by emord** [comment]
I'm ok with 3, but maybe 4 would be better.

what if we had some structure:
```json
{
    "production": {
         "xforms_index": {"settings": {"num_shards": 20}}
     },
    "softlayer": {... },
    ....
}
```
and instead of having the if statement in random pillow mapping, we just grabbed the corresponding index settings in all mapping files and merged the dictionaries?

that's slightly less hacky and still keeps the differences in one place

**Comment by benrudolph** [comment]
yeah that's exactly what i'm thinking. ok cool i'll do that

**Comment by benrudolph** [comment]
there are some gotcha's with that approach. i'm going to bring this up at the scale meeting tomorrow discuss

**Comment by dimagimon** [comment]
There are 5 errors:

* corehq/ex-submodules/pillowtop/es_utils.py, line 74 - F821 undefined name 'index_info'
* corehq/pillows/mappings/case_mapping.py, line 1 - F401 'deepcopy' imported but unused
* corehq/pillows/mappings/case_mapping.py, line 2 - F401 'settings' imported but unused
* corehq/pillows/mappings/xform_mapping.py, line 1 - F401 'deepcopy' imported but unused
* corehq/pillows/mappings/xform_mapping.py, line 2 - F401 'settings' imported but unused

**Comment by benrudolph** [comment]
Ok this went through some churn so probably better to review all at once. However this commit https://github.com/dimagi/commcare-hq/pull/15550/commits/a7531734c46a4e680c81f4a1353855b6d4213b80 is pretty important because it specifies the change we are making to the prod indices

**Review by snopoke — APPROVED** [review]

File: env_settings.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+                            "type": "custom",
+                            "tokenizer": "keyword",
+                            "filter": ["lowercase"]
+                        },
+                    },
+                },
+            },
+        }
+    },
+}
```
Comment (snopoke) [inline env_settings.py]: Need to add ICDS to this as well. With the new infra we'll have 4 ES nodes. Not sure how many shards we want.

**Review by emord — APPROVED** [review]

**Comment by emord** [comment]
@benrudolph do you plan on temporarily disabling shard reallocation while doing this?

**Comment by benrudolph** [comment]
> do you plan on temporarily disabling shard reallocation while doing this?

yes, we may also want to permanently disable that for prod considering the cluster always goes down when that happens.

File: env_settings.py (hunk 2 above)
Comment (benrudolph) [inline env_settings.py]: 👍 i may do that in another PR unless the timing works out well

**Comment by benrudolph** [comment]
https://github.com/dimagi/commcare-hq/pull/15550/commits/13e710d6b0abd74ed2bf9ffb3cbbb04e6afcd786 - @sravfeyn i don't have context on this change, is it fine for me to just have this commit? or do i need to include hq changes?

**Comment by emord** [comment]
I think depending on the answer to https://github.com/dimagi/commcare-hq/pull/14535/files#r109478389 You may be able to not do this

**Comment by sravfeyn** [comment]
The additional change required is to update `CASE_INDEX` per emord's comment.

**Comment by benrudolph** [comment]
Ok this now only will affect the form index

**Comment by emord** [comment]
looks good from my end. Do you want to pull out everything except for the index changes into another PR? might get annoying if there are conflicts in the next 2 weeks while the reindex runs

**Comment by benrudolph** [comment]
good thought (not pressing though, can do that at any time)
